File: LeetCode/Python/solutions/p0023.py
import collections
from typing import List, Optional

from Python.auxiliary.ListNode import ListNode


class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        #
        # Solution v1: Back Track
        #
        # Runtime: 85 ms @ (beats) 99.9%
        # Memory Usage: 18.4 MB @ (beats) 17.41%
        #
        vals = []
        for nodes in lists:
            x = nodes
            while x:
                vals += [x.val]
                x = x.next

        vals = sorted(vals, reverse=True)

        ans = None
        for val in vals:
            ans = ListNode(val, ans)
        return ans

File: LeetCode/Java/src/solutions/p0023_MergekSortedLists/P0023.java
package solutions.p0023_MergekSortedLists;

import auxiliary.ListNode;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class P0023 {
    public ListNode mergeKLists(ListNode[] lists) {
        /*
         * Solution v1: Back Track
         *
         * Runtime: 7 ms @ (beats) 34.43%
         * Memory Usage: 43.8 MB @ (beats) 89.84%
         */
        List<Integer> vals = new ArrayList<>();

        for (ListNode nodes : lists) {
            while (nodes != null) {
                vals.add(nodes.val);
                nodes = nodes.next;
            }
        }

        Collections.sort(vals);

        ListNode ans = null;
        for (int i = vals.size() - 1; i >= 0 ; i--) {
            ans = new ListNode(vals.get(i), ans);
        }

        return ans;
    }
}
